tokio/src/util/linked_list.rs

```diff
@@ -4,95 +4,121 @@
 //! structure's APIs are `unsafe` as they require the caller to ensure the
 //! specified node is actually contained by the list.
 
-use core::cell::UnsafeCell;
-use core::marker::PhantomPinned;
-use core::pin::Pin;
+use core::marker::PhantomData;
 use core::ptr::NonNull;
 
-/// An intrusive linked list of nodes, where each node carries associated data
-/// of type `T`.
+/// An intrusive linked list.
+///
+/// Currently, the list is not emptied on drop. It is the caller's
+/// responsibility to ensure the list is empty before dropping it.
 #[derive(Debug)]
-pub(crate) struct LinkedList<T> {
-    head: Option<NonNull<Entry<T>>>,
-    tail: Option<NonNull<Entry<T>>>,
+pub(crate) struct LinkedList<T: Link> {
+    /// Linked list head
+    head: Option<NonNull<T::Target>>,
+
+    /// Linked list tail
+    tail: Option<NonNull<T::Target>>,
+
+    /// Linker
+    _p: PhantomData<T>,
 }
 
-unsafe impl<T: Send> Send for LinkedList<T> {}
-unsafe impl<T: Sync> Sync for LinkedList<T> {}
+unsafe impl<T: Link> Send for LinkedList<T> where T::Target: Send {}
+unsafe impl<T: Link> Sync for LinkedList<T> where T::Target: Sync {}
+
+/// Defines how a type is tracked within a linked list.
+///
+/// In order to support storing a single type within multiple lists, accessing
+/// the list pointers is decoupled from the entry type.
+///
+/// # Safety
+///
+/// Implementations must guarantee that `Target` types are pinned in memory. In
+/// other words, when a node is inserted, the value will not be moved as long as
+/// it is stored in the list.
+pub(crate) unsafe trait Link {
+    /// Handle to the list entry.
+    ///
+    /// This is usually a pointer-ish type.
+    type Handle;
+
+    /// Node type
+    type Target;
 
-/// A node which carries data of type `T` and is stored in an intrusive list.
+    /// Convert the handle to a raw pointer
+    ///
+    /// Consumes ownership of the handle.
+    fn to_raw(handle: Self::Handle) -> NonNull<Self::Target>;
+
+    /// Convert the raw pointer to a handle
+    fn from_raw(ptr: NonNull<Self::Target>) -> Self::Handle;
```

Comment:
It is safe to call, but not necessarily to implement. This is why the trait itself is `unsafe`.

Reply:
I guess i should check unsafe guidelines again...

Reply:
I'm not 100% sure, but I guess I can change to `unsafe` :shrug:

Reply:
_is_ it safe to call? couldn't a caller pass in a `NonNull` that doesn't have a stable address, because the trait can't really enforce that all `Target`s are immovable?

Reply:
I changed it... but, as of this PR no implementations required unsafe.